### DISCUSSION TITLE
Generating components with cva

### DISCUSSION THREAD

**Discussion by diboune** [joe-bell/cva#49] [Ideas]
Came across the project when trying to migrate off `stitches`, and am loving it so far.
I created a utility function (that looks like stitches) that helps create a react component with the help of `cva`

`Styled.tsx`
```JS
function Styled(htmlTag : keyof JSX.IntrinsicElements, baseClass: string | string[] , options?: any) {

    const Comp = cva(baseClass, options)


    type ComponentProps = JSX.IntrinsicElements['button'] & VariantProps<typeof Comp>
// I know this shouldn't be a "button" but I don't know how to get a type dynamically, my ts is bad

    const ResultComp : FC<ComponentProps> = props => {
        return createElement(htmlTag, {className: Comp(props)}, props.children)
    }
    return ResultComp
}

export default Styled
```

this is how it's used in `index.tsx`
```JS
import './App.css'
import Styled from './styled'

const Button = Styled("button", "button-class", {
  variants: {
    active: {
      true: "button-active",
    }
  }
})

function Index() {
  return (
    <div>
      <Button active>Hey World</Button>
    </div>
  )
}

export default Index
```

This but with better type-safety would be a cool feature, if it's not in the scope of the project can someone help improve this snippet :)

**Comment by joe-bell** [comment]
Nice! Perhaps I should add something to the README, but I _intentionally_ decided not to provide a `styled` API for a few reasons:

1. Polymorphism is tough to type
2. A return to "CSS class-based" components. If you think about components as css classes rather than React.js components, you can:
    1. Use them in any framework
    2. Less JavaScript === smaller bundlesize
    3. Free polymorphism (just apply the CSS class)

**Reply by joe-bell** [reply]
Added to the README

https://github.com/joe-bell/cva#why-dont-you-provide-a-styled-api
